tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -887,7 +889,104 @@ class SPIRVEmitter : public ASTConsumer {
     theBuilder.setInsertPoint(mergeBB);
   }
 
-  void doForStmt(const ForStmt *forStmt) {
+  // TODO: add attributes (roll and unroll)
+  void doWhileStmt(const WhileStmt *whileStmt,
+                   llvm::ArrayRef<const Attr *> attrs = {}) {
+    // While loops are composed of:
+    //   while (<check>)  { <body> }
+    //
+    // SPIR-V requires loops to have a merge basic block as well as a continue
+    // basic block. Even though while loops do not have an explicit continue
+    // block as in for-loops, we still do need to create a continue block.
+    //
+    // Since SPIR-V requires structured control flow, we need two more basic
+    // blocks, <header> and <merge>. <header> is the block before control flow
+    // diverges, and <merge> is the block where control flow subsequently
+    // converges. The <check> block can take the responsibility of the <header>
+    // block. The final CFG should normally be like the following. Exceptions
+    // will occur with non-local exits like loop breaks or early returns.
+    //
+    //            +----------+
+    //            |  header  | <------------------+
+    //            | (check)  |                    |
+    //            +----------+                    |
+    //                 |                          |
+    //         +-------+-------+                  |
+    //         | false         | true             |
+    //         |               v                  |
+    //         |            +------+     +------------------+
+    //         |            | body | --> | continue (no-op) |
+    //         v            +------+     +------------------+
+    //     +-------+
+    //     | merge |
+    //     +-------+
+    //
+    // For more details, see "2.11. Structured Control Flow" in the SPIR-V spec.
+
+    spv::LoopControlMask loopControl = spv::LoopControlMask::MaskNone;
```

Comment:
It's debatable whether we should do the unroll directly in the compiler. I remember that ISVs  do not trust IHVs to do the unrolling so that `Unroll` mask in the SPIR-V code may have no effect at all. (Note that IHVs need to evaluate the SPIR-V and run them in their own backend compiler/interpreter or something in the driver and they may just ignore the mask.)

But not requesting unroll in this change (it is definitely a complicated optimization pass). Having `Unroll` generated into the SPIR-V is definitely better than ignoring the attribute in the source code.

Just wanted to raise the issue to let you know.